consensus/src/payload_generator.rs

```diff
@@ -0,0 +1,367 @@
+    // end_index.
+    start_index: i64,
+    // Hashmap of senders to the number of transactions included in the window for the corresponding
+    // sender.
+    senders_in_window: HashMap<AccountAddress, usize>,
+    // Partially ordered transactions, needs to be updated every time add_transactions is called.
+    txns: Option<Vec<SignedTransaction>>,
+}
+
+impl SenderAwarePayloadState {
+    pub fn new(window_size: usize) -> Self {
+        Self {
+            start_index: -(window_size as i64),
+            senders_in_window: HashMap::new(),
+            txns: Some(Vec::new()),
+        }
+    }
+
+    /// Slides the current window. Essentially, it increments the start_index, end_index and
+    /// updates the senders_in_window map.
+    pub fn add_transaction(&mut self, txn: SignedTransaction) {
+        if self.start_index >= 0 {
+            // if the start_index is negative, then no sender falls out of the window.
+            let sender = self
+                .txns
+                .as_mut()
+                .expect("add transaction called after finalize")
+                .get(self.start_index as usize)
+                .expect("Transaction expected")
+                .sender();
+            self.senders_in_window
+                .entry(sender)
+                .and_modify(|count| *count -= 1);
+        }
+        let count = self
+            .senders_in_window
+            .entry(txn.sender())
+            .or_insert_with(|| 0);
+        *count += 1;
+        self.txns
+            .as_mut()
+            .expect("add transaction called after finalize")
+            .push(txn);
+        self.start_index += 1;
+    }
+
+    pub fn has_conflict_in_window(&self, addr: &AccountAddress) -> bool {
+        self.senders_in_window
+            .get(addr)
+            .map_or(false, |count| *count != 0)
+    }
+
+    pub fn finalize(&mut self) -> Vec<SignedTransaction> {
+        NUM_SENDERS_IN_BLOCK.set(self.senders_in_window.len() as f64);
+        self.txns.take().expect("Finalize already called")
+    }
+}
+
+/// This represents a contiguous chunk of transactions in a block grouped by the same sender.
+struct TransactionsChunkBySender {
+    sender: AccountAddress,
+    transactions: VecDeque<SignedTransaction>,
+}
+
+impl TransactionsChunkBySender {
+    pub fn new(sender: AccountAddress) -> Self {
+        Self {
+            sender,
+            transactions: VecDeque::new(),
+        }
+    }
+
+    pub fn add_transaction(&mut self, txn: SignedTransaction) {
+        self.transactions.push_back(txn);
+    }
+
+    pub fn remove_transaction(&mut self) -> Option<SignedTransaction> {
+        self.transactions.pop_front()
+    }
+
+    pub fn len(&self) -> usize {
+        self.transactions.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.transactions.is_empty()
+    }
+
+    pub fn sender(&self) -> AccountAddress {
+        self.sender
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
can you add these unit tests (because I think there might be some issues/corner cases with the implementation, and it is simpler to confirm/debug this way):
(format - S2_3 - S2 is sender index - second sender, 3 is third transaction from that sender)

test1:
S1_1, S2_1, S1_2, S3_1, S4_1, S5_1 
with conflict_window_size=3, should return (we separate transactions from same sender, even if they are not consecutive):
S1_1, S2_1, S3_1, S4_1, S1_2, S5_1

test2:
S1_1, S1_2, S2_1, S3_1, S3_2,S4_1, S5_1, S6_1
with conflict_window_size=3, should return (each batches are separated from the point they appear on):
S1_1, S2_1, S3_1, S4_1, S1_2, S5_1, S3_2, S6_1

test3:
S1_1, S2_1, S3_1, S3_2
with conflict_window_size=3, should return (keep the order, fairness to early transactions):
S1_1, S2_1, S3_1, S3_2